### PR TITLE
Add `minetest.register_node` and its definition structure

### PR DESCRIPTION
**Pull request by veprogames**
# What does this add?

This adds documentation for `minetest.register_node` and its node definition. It adds the base structure for the associated types like tilespecs, colorspecs and soundspecs, but the documentation for them is not included. The descriptions are taken from lua_api.txt and adjusted.

# Open Questions

* What is the best way to add function signatures, including proper typing?
* Where can default values be added?
* Are there syntax errors I overlooked?

# Goals

Aside from being easily understandable, it should be possible to have a good conversion into language server / intellisense enhancements so you get decent completion and type checking support in common editors like, for example, VSCode(ium) or NeoVim.

### CODE REVIEW

**Comment by appgurueu** [comment]
> What is the best way to add function signatures, including proper typing?

I'm not sure I get the question. We use AsciiDoc tables to document parameters. You're doing it right currently.

> Where can default values be added?

If a function has particularly many defaults, a column may be appropriate. Otherwise I'd just stick it in the field description ("optional, defaults to ...").

> Are there syntax errors I overlooked?

That's the problem with most lean Markup languages: They know no syntax errors :P - if you get the syntax wrong, it'll just be rendered as plain text; preview the document using AsciiDoctor to find these. We can (try to) catch such "syntax errors" during review, but I wouldn't worry about them just yet.

**Comment by veprogames** [comment]
> I'm not sure I get the question. We use AsciiDoc tables to document parameters. You're doing it right currently.

I feel type-function is too generic, I would prefer something in the direction of function(player: PlayerRef, meta: NodeMetaDataRef) so you can see what each parameter is exactly. could also save a little text in the field description. Other technologies that parse the docs can make good use of very concise information. I know Lua has weak typing, but hypothetical linters could still infer outgoing from the docs. I am just not sure how to note it. Maybe `{type-function}(player: {type-playerref}, meta:{type-nodemetadataref})`?

> If a function has particularly many defaults, a column may be appropriate. Otherwise I'd just stick it in the field description ("optional, defaults to ...").

I guess adding a 4th column with the default value, or empty value if required, is a solution. but what to add if I don't know the default value? TODO, WIP, unknown?

> That's the problem with most lean Markup languages: They know no syntax errors :P - if you get the syntax wrong, it'll just be rendered as plain text; preview the document using AsciiDoctor to find these. We can (try to) catch such "syntax errors" during review, but I wouldn't worry about them just yet.

I use a plugin in VSCodium that renders ASCIIDoc previews. Markup stuff wasn't rendered in table cells, but dunno if it's just the plugin or if ASCIIDoc doesn't support it.

**Comment by Minetest-j45** [comment]
I would like to second appguru's thanks for your work @veprogames . However, while reading your comment, I noticed that you mentioned that you took some things from `lua_api.txt`. I thought this might be problematic so I mentioned it to a few members of the team that I am close to, and they told me that my initial understanding was correct.

At the time of writing this comment, this project is licensed under CC-BY 4.0, and `lua_api.txt` is licensed under CC-BY-SA 3.0 (according to @rollerozxa, in #50, however i am not sure of the validity of this claim, since the `LICENSE.txt` in the mt repo doesnt specify the license applying to the documentation, including `lua_api.txt`. Someone more experienced should look into this.)

However, if what @rollerozxa says is true, then copying and changing stuff from `lua_api.txt`, is a derivative, and the license we use isnt compatible with the one the mt docs do and therefore this isnt the legal thing to do.

It seems like the best way to move forward is that this PR waits until we do something about the license incompatibly, as suggested in #50


TL;DR
I know I can tend to be abit repetitive so here is a very brief summary:
The license we use is incompatible with `lua_api.txt`, meaning you cant just copy stuff from it and change it. We are gonna try to change this.


Please take everything I say with a grain, or ideally more, of salt. IANAL

**Comment by appgurueu** [comment]
> I feel type-function is too generic, I would prefer something in the direction of function(player: PlayerRef, meta: NodeMetaDataRef) so you can see what each parameter is exactly. could also save a little text in the field description. Other technologies that parse the docs can make good use of very concise information. I know Lua has weak typing, but hypothetical linters could still infer outgoing from the docs. I am just not sure how to note it. Maybe `{type-function}(player: {type-playerref}, meta:{type-nodemetadataref})`?

Ah, callback functions. I'd use the exact same formatting here, just nesting it. This would give you... a table and some headings in a table. If you think that's too messy, just name it (e.g. MyFancyCallback), then document their signature under a distinct section MyFancyCallback.

**Comment by veprogames** [comment]
I am using this format now: `{type-function}(pos:{class-vector}, node:{class-noderef}, pointed_thing:{type-???})`. If this is wrong, I'd go for nesting, but this format should be parseable and is much shorter.

Haven't modded that much, is pointed_thing a NodeRef?

**Comment by appgurueu** [comment]
> Haven't modded that much, is pointed_thing a NodeRef?

It has its own "pointed thing" type - see https://github.com/minetest/minetest_docs/blob/master/doc/classes/Raycast.adoc#pointed-thing

**Comment by veprogames** [comment]
Alright, the register_node and node definition documentation is roughly done, the other referenced types and classes are out of scope for this PR.

I cannot say if everything is factually correct, though. There are some assumptions, especially with default values.

File: doc/types/NodeDefinition.adoc [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+| `liquidtype` | `{type-string}` | specifies liquid flowing physics
+
+* "none":    no liquid flowing physics
+* "source":  spawns flowing liquid nodes at all 4 sides and below; recommended drawtype: "liquid".
+* "flowing": spawned from source, spawns more flowing liquid nodes around it until `liquid_range` is reached; will drain out without a source; recommended drawtype: "flowingliquid".
+
+NOTE: If it's "source" or "flowing", then the `liquid_alternative_*` fields *must* be specified | `"none"`
+
+| `liquid_alternative_flowing` | `{type-string}` | These fields may contain node names that represent the flowing version (`liquid_alternative_flowing`) and source version (`liquid_alternative_source`) of a liquid.
```
Comment (appgurueu) [inline doc/types/NodeDefinition.adoc:72]: These fields? You've only written the flowing one in the column...

File: doc/types/NodeDefinition.adoc [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+* false: No liquid movement physics apply.
+* true: Enables liquid movement physics. Enables things like ability to "swim" up/down, sinking slowly if not moving, smoother speed change when falling into, etc. The `movement_liquid_*` settings apply.
+* nil: Will be treated as true if `liquidtype ~= "none"` and as false otherwise. | `nil`
+
+| `liquid_range` | `{type-number}` | Maximum distance that flowing liquid nodes can spread around source on flat land (max. 8). If this is set to 0, liquid flow will be disabled. | `8`
+
+| `leveled` | `{type-number}` | This only applies to nodes that have the "nodebox" drawtype set to '"leveled"'. This allows defining the nodebox height without using param2. The nodebox height is 'leveled' / 64 nodes. The maximum value of 'leveled' is `leveled_max`. | `0`
+| `leveled_max` | `{type-number}` | Maximum value for `leveled` (0-127), enforced in `minetest.set_node_level` and `minetest.add_node_level`. Values above 124 might causes collision detection issues. | `127`
+
+| `drowning` | `{type-number}` | Player will take this amount of damage per second if no bubbles are left. | `0`
```
Comment (appgurueu) [inline doc/types/NodeDefinition.adoc:109]: (if set to 0, player won't lose air in the node)

File: doc/types/NodeDefinition.adoc [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+Example: `{"group:fence", "default:wood"}` or `"default:stone"` | `{}`
+| `connect_sides` | `{type-table}` | Tells connected nodebox nodes to connect only to these sides of this node. Possible: `"top"`, `"bottom"`, `"front"`, `"left"`, `"back"`, `"right"`. | `{}`
+
+| `mesh` | `{type-string}` | File name of mesh when using the `"mesh"`` drawtype. | `""`
+
+| `selection_box` | `{type-table}` | Custom selection box definition. Multiple boxes can be defined. If `"nodebox"` drawtype is used and `selection_box` is `nil`, then `node_box` definition is used for the selection box. See [Node boxes] for possibilities. | `nil`
+
+| `collision_box` | `{type-table}` | Custom collision box definition. Multiple boxes can be defined. If `"nodebox"` drawtype is used and `collision_box` is `nil`, then `node_box definition` is used for the collision box. See [Node boxes] for possibilities. | `nil`
+
+// is this still needed? (It's been 11 years)
```
Comment (appgurueu) [inline doc/types/NodeDefinition.adoc:123]: I'd say in general we don't have to document legacy APIs...

File: doc/types/NodeDefinition.adoc [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+| `legacy_facedir_simple` | `{type-bool}` | Support maps made in and before January 2012. | `false`
+| `legacy_wallmounted` | `{type-bool}` | Support maps made in and before January 2012. | `false`
+
+| `waving` | `{type-number}` | This is valid for drawtypes: `"mesh"`, `"nodebox"`, `"plantlike"`, `"allfaces_optional"`, `"liquid"` and `"flowingliquid"`. The following values can be set:
+* 1 - wave node like plants (node top moves side-to-side, bottom is fixed)
```
Comment (appgurueu) [inline doc/types/NodeDefinition.adoc:128]: (and 0 - don't wave)

File: doc/types/NodeDefinition.adoc [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+* `"allfaces_optional"` drawtype can only wave like leaves.
+* `"liquid"`, `"flowingliquid"` drawtypes can only wave like liquids. | `0`
+
+| `sounds` | `{type-nodesoundspec}` | Definition of node sounds to be played at various events. See {type-nodesoundspec}. | `nil`
+
+| `drop` | `{type-string} or {type-nodedropspec}` | By default, the dropped item is the node itself. If a {type-string} is given, `drop` will be dropped (for example: `"default:stone"`). Using a table allows multiple items, drop chances and item filtering, see {type-nodedropspec}. | `""`
+
+| `mod_origin` | `{type-string}` | stores which mod actually registered a node. If the source could not be determined, it contains `"??"`. This is useful for getting which mod truly registered something.
+
+Example: if a node is registered as `":othermodname:nodename"`, nodename will show `"othermodname"`, but mod_origin will say `"modname"`. | `"<modname>"`
```
Comment (appgurueu) [inline doc/types/NodeDefinition.adoc:144]: could reference `minetest.get_current_modname()` for the default ;)

File: doc/types/NodeDefinition.adoc [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+| `after_dig_node` | `{type-function}(pos:{class-vector}, oldnode:{class-noderef}, oldmetadata:{type-table}, digger:{class-playerref})` | Called after destructing the node when the node was dug using `minetest.node_dig` or `minetest.dig_node`. | `nil`
+
+| `can_dig` | `{type-function}(pos:{class-vector}, player:{class-playerref}\|{type-nil})` | If the node shall be diggable, this functions should return true, false otherwise. | `nil`
+
+| `on_dig` | `{type-function}(pos:{class-vector}, node:{class-noderef}, pointed_thing:{type-pointedthing})` | Called when the node was dug. By default, this checks privileges (if the node can be dug). If the item used to dig this node is a tool, the tools durability will be deducted. Return true if the node was dug successfully, false otherwise. | `nil`
```
Comment (appgurueu) [inline doc/types/NodeDefinition.adoc:166]: Default is not `nil` but rather `minetest.node_dig`. I feel it would make sense to move this above the corresponding `after_*` callback to make this clear.

File: doc/types/NodeDefinition.adoc [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+
+| `on_flood` | `{type-function}(pos:{class-vector}, oldnode:{class-noderef}, newnode:{class-noderef})` | Called when a liquid (`newnode`) is about to flood `oldnode`, if it has `floodable = true` in the {type-nodedefinition}. This function is not called for bulk node placement (i.e. schematics and VoxelManip) or air nodes. If this function returns true, the node is not flooded, but `on_flood` will most likely be called over and over again every liquid update interval.
+
+WARNING: making a liquid node 'floodable' will cause problems. | `nil`
+
+| `on_punch` | `{type-function}(pos:{class-vector}, node:{class-noderef}, puncher:{class-objectref}, pointed_thing:{type-pointedthing})` | Called when `puncher` punches the `node` at `pos`. By default, `minetest.register_on_punchnode` callbacks are called. | `nil`
```
Comment (appgurueu) [inline doc/types/NodeDefinition.adoc:172]: Default is `minetest.node_punch` here.

File: doc/types/NodeDefinition.adoc [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+
+| `on_receive_fields` | `{type-function}(pos:{type-vector}, formname:{type-string}, fields:{type-table}, sender:{class-playerref})` | Called when a Formspec (e.g. sign text input) returns data. See `minetest.register_on_player_receive_fields` for more info.
```
Comment (appgurueu) [inline doc/types/NodeDefinition.adoc:174]: Called when *the formspec* associated with a node of this type through the `formspec` node meta field (you could reference the docs there) was submitted.

File: doc/types/NodeDefinition.adoc [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+NOTE: the `fields` are supplied in a key-value format: `fields = {name1 = value1, name2 = value2, ...}` | `nil`
```
Comment (appgurueu) [inline doc/types/NodeDefinition.adoc:175]: This would be the scope of formspec docs (I think).

File: doc/types/NodeDefinition.adoc [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+
+| `on_rightclick` | `{type-function}(pos:{class-vector}, node:{class-noderef}, clicker:{class-objectref}, itemstack:{class-itemstack}, pointed_thing:{type-pointedthing})` | Called when `clicker` used the 'place/build' key (not necessarily an actual rightclick) while pointing at the `node` at pos with 'node'. `itemstack` will hold the wielded item of the `clicker`. This function shall return the leftover itemstack.
+
+NOTE: pointed_thing can be nil, if a mod calls this function. This function does not get triggered by clients running version 0.4.16 or earlier if the `"formspec"` node metadata field is set. | `nil`
```
Comment (appgurueu) [inline doc/types/NodeDefinition.adoc:179]: For the future: You definitely don't have to document 0.4.x, it's one major version behind.

File: doc/types/NodeDefinition.adoc [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+
+| `allow_metadata_inventory_put` | `{type-function}(pos:{class-vector}, listname:{type-string}, index:{type-number}, stack:{class-itemstack}, player:{class-playerref})` | Called when a player wants to put something into the inventory. This function should return the number of items allowed to put.
+
+NOTE: If `-1` is returned, the item count in `player`s inventory won't be modified and the put will be allowed. | `nil`
+
+| `allow_metadata_inventory_take` | `{type-function}(pos:{class-vector}, listname:{type-string}, index:{type-number}, stack:{class-itemstack}, player:{class-playerref})` | Called when a player wants to take something out of the inventory. This function should return the number of items `player` is allowed to take.
+
+NOTE: If `-1` is returned, the item count in `player`s inventory won't be modified and the put will be allowed. | `nil`
```
Comment (appgurueu) [inline doc/types/NodeDefinition.adoc:194]: (use this for creative inventories)

**Review by appgurueu — COMMENTED** [review]
Looks fine overall, just some minor things. I noticed but didn't raise a couple typos; we'll edit those out in a final pass.

File: doc/types/NodeDefinition.adoc (hunk 9 above)
Comment (veprogames) [reply doc/types/NodeDefinition.adoc:175]: I'd like to keep the quick hint about the table data structure here.

File: doc/types/NodeDefinition.adoc (hunk 10 above)
Comment (veprogames) [reply doc/types/NodeDefinition.adoc:179]: Then I am also going to remove the `legacy_` properties too ("maps created before on on January 2012")

Comment (appgurueu) [reply doc/types/NodeDefinition.adoc]: Yes, that's fine.

File: doc/types/NodeDefinition.adoc (hunk 9 above)
Comment (appgurueu) [reply doc/types/NodeDefinition.adoc:173]: Sure, that's fine.

**Comment by veprogames** [comment]
Just saw I overlooked some review bits, will do them later

**Comment by veprogames** [comment]
I have done the remaining bits now

**Comment by appgurueu** [comment]
Something I missed: You have to use `a` before a table cell to make AsciiDoc in it render correctly. See e.g. [this commit](https://github.com/minetest/minetest_docs/commit/b47713e48a231943d080aa857a57d99a1d3237e4).

**Comment by veprogames** [comment]
should be an `a` everywhere needed now, also added more links (to empty docs, they are out of scope)

**Review by appgurueu — APPROVED** [review]
Content-wise fine.

Still waiting on the license issue @minetest/docs

**Comment by appgurueu** [comment]
Merged manually via https://github.com/minetest/minetest_docs/commit/be5d9bc89dcc287d0140c73738c61e9645e40aa4 (squash and merge is currently disabled).
